### PR TITLE
Add checks for existing callable at compilation

### LINKED ISSUE #3450: Twig assumes static callable if it does not exist when registering a filter
If I register a class method as a filter callable and it does not exist it is treated as being static. Afterwards if I define the callable as an instance method it is called statically. 

On php 8 this is a fatal error in 7.4 it is called nonetheless and goes through if the method does not use $this.

After the compiled template is removed everything works as expected.

[bug.zip](https://github.com/twigphp/Twig/files/5617282/bug.zip)

### PR DESCRIPTION
**Pull request by ghost**
Fixes #3450 

Currently Twig does not check if the callable passed to CallExpression is actually a callable.
This pull request adds does checks.

If it is a function it check if it exists. If it is an array indicating a class method, then it also checks for the presence of __call and __callStatic methods.
If the callable does not exist a \LogicException is thrown. (I don't know if this is the appropriate Exception).

I think Twig should check for the existence of callables if it can otherwise it can be quite cumbersome to debug when using caching, since it is possible a wrong path will be compiled. (This happened to me.)

Some unrelated Function/Filter tests broke because they were passing undefined callables so I fixed them. However, the intent was clear to pass an existing callable.

### CODE REVIEW

File: src/Node/Expression/CallExpression.php [hunk 1]
```diff
@@ -293,11 +296,16 @@ private function reflectCallable($callable)
         }
 
         if (\is_array($callable)) {
-            if (!method_exists($callable[0], $callable[1])) {
-                // __call()
+            list($class, $method) = $callable;
```
Comment (stof) [inline src/Node/Expression/CallExpression.php]: The naming `$class` is confusing here, as this is really `$objectOrClass`

Comment (ghost) [reply src/Node/Expression/CallExpression.php]: Is it ok that I destructed the callable?

File: src/Node/Expression/CallExpression.php [hunk 2]
```diff
@@ -306,12 +314,18 @@ private function reflectCallable($callable)
             $class = substr($callable, 0, $pos);
             $method = substr($callable, $pos + 2);
             if (!method_exists($class, $method)) {
+                if (!method_exists($class, '__callStatic')) {
```
Comment (ghost) [inline src/Node/Expression/CallExpression.php:317]: Should this also check for __call? We allow [FQCN, 'nonStaticMethod'] for instance methods even though that is not a legal callable. Is 'FQCN::nonStaticMethod' also legal in context of Twig?

Comment (ghost) [reply src/Node/Expression/CallExpression.php:317]: @stof What do you think?

**Comment by ghost** [comment]
@fabpot do you have an opinion on this pull request?

**Comment by ghost** [comment]
Sorry for pinging again. Is there anything else I have to do to get this PR merged? Or are these checks not desired? @stof  @fabpot

**Comment by ghost** [comment]
@stof @fabpot  do I have to do anything else?

**Comment by nlemoine** [comment]
I'm having a similar issue here. 

```php
class SomeExtension extends AbstractExtension
{
    private array $filters = [
         'filter1',
         'filter2',
         'filter3',
         // and so on...
    ];

    public function getFilters(): array
    {
        $filters = [];
        foreach ($this->filters as $filter) {
            $filters[] = new TwigFilter(
                $filter,
                // ["SomeRuntime::$filter"], // Don't work
                // [SomeRuntime::class, $filter], // Don't work
                ['is_variadic' => true]
            );
        }

        return $filters;
    }
}

class SomeRuntime implements RuntimeExtensionInterface
{
    private Service $service;

    public function __construct(Service $service)
    {
        $this->service = $service;
    }

    public function __call($name, array $arguments = [])
    {
        $object = $this->service->create($arguments[0]);
        return $object->$name(...$arguments)
    }
}

```

Could you give this PR/issue some a second thought?

**Pull request closed without merging**